common/http/platform/beast/http.cpp

```diff
@@ -118,6 +118,107 @@ class BodyAsyncReader : virtual public io::AsyncReader {
 	friend class Server;
 };
 
+template <typename StreamType>
+class RawSocket : virtual public io::AsyncReadWriter {
+public:
+	RawSocket(shared_ptr<StreamType> stream, shared_ptr<beast::flat_buffer> buffered) :
+		destroying_ {make_shared<bool>(false)},
+		stream_ {stream},
+		buffered_ {buffered} {
+		// If there are no buffered bytes, then we don't need it.
+		if (buffered_ && buffered_->size() == 0) {
+			buffered_.reset();
+		}
+	}
+
+	~RawSocket() {
+		*destroying_ = true;
+		Cancel();
+	}
+
+	error::Error AsyncRead(
+		vector<uint8_t>::iterator start,
+		vector<uint8_t>::iterator end,
+		io::AsyncIoHandler handler) override {
+		// If we have prebuffered bytes, which can happen if the HTTP parser read the
+		// header and parts of the body in one block, return those first.
+		if (buffered_) {
+			size_t to_copy = min(static_cast<size_t>(end - start), buffered_->size());
```

Comment:
Could split it out for readability, and call it something like `emptyBuffer` or similar?